subprojects/core/src/main/java/org/gradle/execution/plan/BuildInputHierarchy.java

```diff
@@ -0,0 +1,122 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.execution.plan;
+
+import org.gradle.api.file.FileTreeElement;
+import org.gradle.api.specs.Spec;
+import org.gradle.execution.plan.ValuedVfsHierarchy.ValueVisitor;
+import org.gradle.internal.collect.PersistentList;
+import org.gradle.internal.file.Stat;
+import org.gradle.internal.snapshot.CaseSensitivity;
+import org.gradle.internal.snapshot.VfsRelativePath;
+
+import java.io.File;
+import java.util.function.Supplier;
+
+/**
+ * Allows recording and querying the input locations of a build.
+ */
+public class BuildInputHierarchy {
+    private volatile ValuedVfsHierarchy<InputDeclaration> root;
+    private final SingleFileTreeElementMatcher matcher;
+
+    public BuildInputHierarchy(CaseSensitivity caseSensitivity, Stat stat) {
+        this.root = ValuedVfsHierarchy.emptyHierarchy(caseSensitivity);
+        this.matcher = new SingleFileTreeElementMatcher(stat);
+    }
+
+    /**
+     * Returns whether the given location is an input to the build.
+     */
+    public boolean isInput(String location) {
+        InputDeclarationVisitor visitor = new InputDeclarationVisitor();
+        root.visitValues(location, visitor);
+        return visitor.isInput();
+    }
+
+    public boolean isEmpty() {
+        return root.isEmpty();
+    }
+
+    /**
+     * Records that some locations are an input to the build.
+     */
+    public synchronized void recordInputs(Iterable<String> inputLocations) {
+        for (String location : inputLocations) {
+            VfsRelativePath relativePath = VfsRelativePath.of(location);
+            root = root.recordValue(relativePath, ALL_CHILDREN_ARE_INPUTS);
+        }
+    }
+
+    /**
+     * Records that a fileTreeRoot with a filter is an input to the build.
+     *
+     * Only children of the fileTreeRoot that match the filter are considered inputs.
+     */
+    public synchronized void recordFilteredInput(String fileTreeRoot, Spec<FileTreeElement> filter) {
+        VfsRelativePath relativePath = VfsRelativePath.of(fileTreeRoot);
+        root = root.recordValue(relativePath, new FilteredInputDeclaration(filter));
+    }
+
+    private static class InputDeclarationVisitor implements ValueVisitor<InputDeclaration> {
+        private boolean input = false;
+
+        private void foundInput() {
+            this.input = true;
+        }
+
+        @Override
+        public void visitExact(InputDeclaration value) {
+            foundInput();
+        }
+
+        @Override
+        public void visitAncestor(InputDeclaration value, VfsRelativePath pathToVisitedLocation) {
+            if (value.isInput(pathToVisitedLocation)) {
```

Comment:
This took me a while to understand. I think with a few renames we could make it easier to grasp. WDYT of:

```suggestion
            if (ancestor.contains(pathFromAncestor)) {
```

`containsInput()` could be another option.

Reply:
Renamed to `contains`.